File: design/insert_delete_getrandom.kt
fun main() {
    val r = RandomizedSet()

    println(r.remove(4)) // false
    println(r.insert(4)) // true
    println(r.insert(4)) // false
    println(r.insert(9)) // true
    println(r.remove(9)) // true
    println(r.getRandom())
}


class RandomizedSet() {

    /** Initialize your data structure here. */
    val values = mutableMapOf<Int, Int>()
    val random = mutableListOf<Int>()


    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    fun insert(`val`: Int): Boolean {
        // O(1) time, O(n) space

        // val already in the set, so return false
        if (`val` in values) return false

        // map val to the last index in random
        values[`val`] = random.size
        // append val to random
        random.add(`val`)

        return true
    }

    /** Removes a value from the set. Returns true if the set contained the specified element. */
    fun remove(`val`: Int): Boolean {
        // O(1) time, O(n) space

        // val not in set, so return false
        if (`val` !in values) return false

        // get the index into random for this value
        val rIndex = values[`val`]!!
        // remove val from the set
        values.remove(`val`)

        // replace the value in random with the last value
        random[rIndex] = random[random.size-1]
        // remove the last value from random
        random.removeAt(random.size-1)

        // if the removed value wasn't the last one in random,
        // update the map for the newly moved value in random
        if (rIndex < random.size ) {
            values[random[rIndex]] = rIndex
        }

        return true
    }

    /** Get a random element from the set. */
    fun getRandom(): Int {
        // O(1) time, O(n) space

        // throw if set is empty
        if (random.size <= 0) {
            throw Exception("Empty Set")
        }

        return random[(0 until random.size).random()]
    }

}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = RandomizedSet()
 * var param_1 = obj.insert(`val`)
 * var param_2 = obj.remove(`val`)
 * var param_3 = obj.getRandom()
 */

File: design/insert_delete_getrandom.py
#! /usr/bin/env python3

from random import randint


class RandomizedSet:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.ht = {}
        self.random_access = []

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already
        contain the specified element.
        """
        if val in self.ht:
            return False

        index = len(self.random_access)
        self.random_access.append(val)
        self.ht[val] = index

        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the
        specified element.
        """
        if val not in self.ht:
            return False

        index = self.ht[val]
        self.random_access[index] = self.random_access[-1]

        if index < len(self.random_access):
            self.ht[self.random_access[index]] = index

        del(self.ht[val])
        self.random_access.pop()
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return self.random_access[randint(0, len(self.random_access)-1)]


if __name__ == "__main__":
    rs = RandomizedSet()
    print(rs.insert(0))  # true
    print(rs.insert(1))  # true
    print(rs.remove(0))  # true
    print(rs.insert(2))  # true
    print(rs.remove(1))  # true
